### ISSUE TITLE
Consider the impact of out-of-order ingestion limitation and users coming from og-pyroscope.

### ISSUE THREAD

**Issue by simonswine** [grafana/pyroscope#2198]
We are enforcing a strict time series order policy for a series (at least within a single head block) on the ingesters.

This means let's say someone runs a java process with `n > 1 replicas` with the exact same config:

```
  profiler.Start(profiler.Config{
    ApplicationName: "my.app",
    ServerAddress:   "http://pyroscope:4040",
  })
```

They most likely will compete on who sends the newest profiles. As there is no differentiator label between the different pods.

So this will cause a lot out of order errors and will also confuse the user why that is.

I think we need to reconsider that strict ording and maybe should allow a wider time window being sent and not strictly in order.

**Comment by simonswine** [comment]
cc @cyriltovena @kolesnikovae @korniltsev

**Comment by petethepig** [comment]
Ah, that explains those out of order errors, thanks for explaining!

**Comment by kolesnikovae** [comment]
As far I as I can judge, the only actual reason OOO profiles are rejected is the fact that we need to calculate diff for delta profiles (in ingesters), and thus we have to ensure that the profiles are consecutive.

I think we could try to prohibit cumulative profiles (that need calculating the diff, such as Go heap) in ingesters completely:
- In push mode, delta should be calculated in the client (e.g. godeltaprof).
- In pull mode, we can distinguish targets easily and delta should be calculated in scrappers:
  - Or, better, we offer profile exporter that calculates the profile diff, so that scrappers do nothing (e.g godeltaprof).
  - If the target covers multiple instances (e.g behind a load balancer), delta thing does not work either way, so we can only conclude that this scenario is not supported.

That would allow us to allow OOO ingestion. Another point is that we may want to keep some configurable _ingestion window_ to avoid blocks with too wide time range.

The very fact that profiles are not ordered by time within its series when they reside in memory should not be very harmful and can be mitigated in various ways (e.g partial sorting). Of course, profiles must be ordered when written to persistent chunk/block.

Moreover, if we want to support FaaS integrations (e.g AWS Lambda), OOO profiles must be handled.

**Comment by kolesnikovae** [comment]
Solved in #2452
